pkg/controllers/cassandra/role/control_test.go

```diff
@@ -1,40 +1,93 @@
 package role_test
 
 import (
-	"fmt"
 	"testing"
 
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/jetstack/navigator/internal/test/unit/framework"
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
 	"github.com/jetstack/navigator/pkg/controllers/cassandra/role"
 	casstesting "github.com/jetstack/navigator/pkg/controllers/cassandra/testing"
 )
 
 func TestRoleSync(t *testing.T) {
-	t.Run(
-		"role missing",
-		func(t *testing.T) {
-			f := casstesting.NewFixture(t)
-			f.Run()
-			f.AssertRolesLength(1)
+	cluster1 := casstesting.ClusterForTest()
+	role1 := role.RoleForCluster(cluster1)
+
+	type testT struct {
+		kubeObjects        []runtime.Object
+		cluster            *v1alpha1.CassandraCluster
+		fixtureManipulator func(*testing.T, *framework.StateFixture)
+		assertions         func(*testing.T, *controllers.State, testT)
+		expectErr          bool
+	}
+	tests := map[string]testT{
+		"create if not listed": {
+			cluster: cluster1,
+			assertions: func(t *testing.T, state *controllers.State, test testT) {
+				expectedObject := role1
+				_, err := state.Clientset.
+					RbacV1beta1().
+					Roles(expectedObject.Namespace).
+					Get(expectedObject.Name, v1.GetOptions{})
+				if err != nil {
+					t.Error(err)
+				}
+			},
 		},
-	)
-	t.Run(
-		"role exists",
-		func(t *testing.T) {
-			f := casstesting.NewFixture(t)
-			existingRole := role.RoleForCluster(f.Cluster)
-			f.AddObjectK(existingRole)
-			f.Run()
-			f.AssertRolesLength(1)
+
+		"no error if already listed": {
+			kubeObjects: []runtime.Object{role1},
+			cluster:     cluster1,
 		},
-	)
-	t.Run(
-		"sync fails",
-		func(t *testing.T) {
-			f := casstesting.NewFixture(t)
-			f.RoleControl = &casstesting.FakeControl{
-				SyncError: fmt.Errorf("simulated sync error"),
-			}
-			f.RunExpectError()
+		"no error if not yet listed": {
```

Comment:
What test case is this checking? Where the resource exists in the client, but hasn't been seen by the lister yet?

Seems to me like we *would* expect to get an error in that case, as we'd attempt to create the resource (and then it fails as it already exists). Then, the next sync will occur after a period of time long enough for the lister to be up to date.